`InlineArray`'s `Equals` and `GetHashCode` throw

AaronRobinsonMSFT opened the pull request:
The default `Equals()` and `GetHashCode()` throw
for types marked with `InlineArrayAttribute`.

Fixes https://github.com/dotnet/runtime/issues/96943

See documentation update at https://github.com/dotnet/dotnet-api-docs/pull/10036

AaronRobinsonMSFT committed:
InlineArray's Equals and GetHashCode throw

The default Equals() and GetHashCode() throw
for types marked with InlineArrayAttribute.

jkotas commented on src/coreclr/nativeaot/System.Private.CoreLib/src/Internal/Runtime/CompilerHelpers/ThrowHelpers.cs:
```diff
@@ -132,5 +132,10 @@ public static void ThrowArgumentOutOfRangeException()
+
+        public static void ThrowInvalidOperationInlineArrayEqualsGetHashCode()
+        {
+            throw new InvalidOperationException(SR.InvalidOperation_InlineArrayEqualsGetHashCode);
```
Re-reading the [guidelines](https://learn.microsoft.com/en-us/dotnet/fundamentals/runtime-libraries/system-notsupportedexception#related-exception-types) about InvalidOperationException vs. NotSupportedException, I think this should be NotSupportedException.

AaronRobinsonMSFT committed:
Switch to NotSupportedException.

AaronRobinsonMSFT committed:
Change name of local.

MichalStrehovsky commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -1624,7 +1624,8 @@ BOOL CanCompareBitsOrUseFastGetHashCode(MethodTable* mt)
     if (mt->ContainsPointers()
-        || mt->IsNotTightlyPacked())
+        || mt->IsNotTightlyPacked()
+        || mt->GetClass()->IsInlineArray())
```
I _think_ this might need a matching change on native AOT side (either in RequiresValueTypeGetFieldHelperMethod or `ComparerIntrinsics.CanCompareValueTypeBits`), otherwise we'll have different behavior for InlineArray of size 1 (with e.g. Int32 field) since those typically don't have gaps in the layout and are eligible to be compared by hashing/comparing bits.

Or we could say that InlineArray of size 1 that doesn't contain pointers and is tightly packed has a working Equals/GetHashCode and drop the addition here.

AaronRobinsonMSFT replied:
Indeed. Looks like I'm missing a test case here.

AaronRobinsonMSFT replied:
@MichalStrehovsky Thanks for noting this. Added test and sure enough a check was missing.

AaronRobinsonMSFT committed:
Review feedback.

Add test for types where their layout is the same
whether InlineArray is provided or not.

AaronRobinsonMSFT committed:
Merge remote-tracking branch 'upstream/main' into runtime_96943

tannergooding commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -1624,7 +1624,8 @@ BOOL CanCompareBitsOrUseFastGetHashCode(MethodTable* mt)
+        || mt->IsNotTightlyPacked()
+        || mt->GetClass()->IsInlineArray())
     {
         mt->SetHasCheckedCanCompareBitsOrUseFastGetHashCode();
```
This is an unfortunate consequence of throwing.

We'd otherwise be able to treat `InlineArray` of primitive types as bitwise equatable and allow them to participate in some of the vectorization speedups.

It would be nice if there was a better way to allow users to override the default while still opting their types into bitwise equality support.

AaronRobinsonMSFT replied:
I think this is a point in time issue. This is merely the simplest way to block the scenario. It is entirely possible how we block this is changed in the future in a way that permits vectorization speed ups. Filed an issue: https://github.com/dotnet/runtime/issues/103712.

tannergooding replied:
👍, was more just calling it out as something for us to make explicitly known, not to block the change or anything like that.

It's just a broader issue where anytime the user overrides default behavior, we lose the ability to optimize things as we can no longer assume the implementation (even if the user implementation does exactly what we'd want).
